pkg/pbq/pbq.go

```diff
@@ -0,0 +1,100 @@
+package pbq
+
+import (
+	"context"
+	"errors"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/pbq/store"
+	"go.uber.org/zap"
+)
+
+var COBErr error = errors.New("error while writing to pbq, pbq is closed")
+
+// PBQ Buffer queue which is backed with a persisted store, each partition
+// will have a PBQ associated with it
+type PBQ struct {
+	store       store.Store
+	output      chan *isb.Message
+	cob         bool // cob to avoid panic in case writes happen after close of book
+	partitionID string
+	options     *options
+	manager     *Manager
+	log         *zap.SugaredLogger
+}
+
+var _ ReadWriteCloser = (*PBQ)(nil)
+
+// Write writes message to pbq and persistent store
+func (p *PBQ) Write(ctx context.Context, message *isb.Message) error {
+	// if cob we should return
+	if p.cob {
+		p.log.Errorw("failed to write message to pbq, pbq is closed", zap.Any("partitionID", p.partitionID), zap.Any("header", message.Header))
+		return COBErr
+	}
+	var writeErr error
+	// we need context to get out of blocking write
+	select {
+	case p.output <- message:
+		// this store.Write is an `inSync` flush (if need be). The performance will be very bad but the system is correct.
+		// TODO: shortly in the near future we will move to async writes.
+		writeErr = p.store.Write(message)
+	case <-ctx.Done():
```

Comment:
I think this `case` is not needed, also there's no need to use `select`.

Should not `close(p.output)` there as well.

Reply:
we cannot close it here because it will be signaling a close-of-book. ctx.Done was added there if avoid a holdup during the shutdown (in case there is a back-pressure and we are not able to write to the forwarding channel).

Reply:
When using `select`, if it's already in a `case` execution (let's say, here in `p.store.Write()`), it will not respond to another `case` (`ctx.Done()`). If `p.store.Write()` is done with an error return, it breaks out the `select` directly, so there's no need to have a `case <- ctx.Done()` here.

Reply:
If we want to address the store writing back-pressure issue, we should introduce a `context` into `p.store.Write()`.

Reply:
select is to avoid infinite blocking on the output channel

Reply:
the `ctx` in only for the channel, ATM we are not worried about the write to the store in a blocked state. we will have to revisit when we have a proper disk store. there is already a TODO for this.

Reply:
> select is to avoid infinite blocking on the output channel

See my comments above, that will not work.

Reply:
yeah, I know it won't work on the store, but as mentioned by @vigith we are not worried about the writing to the store getting blocked, context and select are to avoid blocked write on the output channel

Reply:
Makes sense!

One last question - do we really want to `close(p.output)` here? Once it's closed, how to continue writing to it?

Reply:
we are closing the channel on ctx.Done(), so there will not be any writes after that right?

Reply:
This is to assume the `write()` either succeed or block until exit (and no way to continue). Imagine the invoker wants to call `Write()` with a timeout in the context, in this case there's no way to let him/her retry after timeout.

Reply:
Yeah, `Write()` with timeout is not supported, I don't think we should. @vigith @ashwinidulams WDYT?

Reply:
we shouldn't support writes with timeout. actually there is no difference between a blocking write and a write with timeout in this context since if the write doesn't go through or is taking long time, then it means there is some back pressure. with timeout, its no different from a blocking write on the left side of PBQ.

we will also offer option for pipeline developers to pick async flush option which will eliminate the need for blocking.

Reply:
Ok.